fmpz_poly/div_series_basecase.c

```diff
@@ -31,59 +43,147 @@ _fmpz_poly_div_series_basecase(fmpz * Q, const fmpz * A, slong Alen,
                 _fmpz_vec_set(Q, A, Alen);
             else
                 _fmpz_vec_neg(Q, A, Alen);
-        } else
+        }
+        else
         {
-            fmpz_init(r);
-
+            slong i;
             for (i = 0; i < Alen; i++)
-            {   
-                fmpz_fdiv_qr(Q + i, r, A + i, B + 0);
-
-                if (!fmpz_is_zero(r))
-                {
-                    fmpz_clear(r);
-
-                    flint_printf("Not an exact division\n");
-                    flint_abort();
-                }
-            }
-
-            fmpz_clear(r);
+                fmpz_divexact_checked(Q + i, A + i, B);
         }
 
         _fmpz_vec_zero(Q + Alen, n - Alen);
     }
+    else if (Alen == 1 && fmpz_is_pm1(B))
+    {
+        _fmpz_poly_inv_series_basecase(Q, B, Blen, n);
+        if (!fmpz_is_one(A))
+            _fmpz_vec_scalar_mul_fmpz(Q, Q, n, A);
+    }
     else
     {
-        slong i, j;
+        slong i, j, nsmall;
+        char *Bbits;
+        slong b, bits, Qbits;
+        TMP_INIT;
+
+        TMP_START;
 
         if (fmpz_is_pm1(B))
         {
             if (fmpz_is_one(B))
                 fmpz_set(Q, A);
             else
                 fmpz_neg(Q, A);
-        } else
+        }
+        else
         {
-            fmpz_init(r);
+            fmpz_divexact_checked(Q, A, B);
+        }
 
-            fmpz_fdiv_qr(Q + 0, r, A + 0, B + 0);
+        /* Bbits[i] = max(bits(B[0]), ..., bits(B[i])), as long as coeffs are small */
+        Bbits = TMP_ALLOC(Blen);
+        Bbits[0] = fmpz_bits(B);
 
-            if (!fmpz_is_zero(r))
-            {
-                fmpz_clear(r);
+        /* Maximum bits of all Q coefficients encountered so far */
+        Qbits = fmpz_bits(Q);
 
-                flint_printf("Not an exact division\n");
-                flint_abort();
-            }
+        /* We have small coefficients for i < nsmall */
```

Comment:
What is i here?

Reply:
The loop index.

Reply:
Oh, it really is i below. Ok, yeah this is fine.